### PR TITLE
ASCollection: implement IEnumerable

### LINKED ISSUE #43: ASCollection: implement IEnumerable<T>
Implement `IEnumerable<T>` to iterate through `Items`. This will allow ForEach over a collection.

### PR DESCRIPTION
**Pull request by CreggHancock**
Resolves #43 

It does not implement the generic `IEnumerable<T>` as `OrderedCollection` is no longer generic. If we still desire it to implement a generic IEnumerable then we'll have to revert or partially revert [this](https://github.com/warriordog/ActivityPubSharp/commit/b4934308bb097ff4989832c9b9953431486e473c), but I think implementing as `IEnumerable<Linkable<ASObject>>` should hopefully not be a problem since that's what the Entity's LinkableList is essentially doing.

### CODE REVIEW

File: ActivityPub.Types/AS/Collection/ASCollection.cs [hunk 1]
```diff
@@ -98,6 +99,32 @@ public int TotalItems
     public bool HasItems => Items?.Any() == true;
 
     public static implicit operator ASCollection(List<ASObject> collection) => new() { Items = new LinkableList<ASObject>(collection) };
+
+    public IEnumerator<Linkable<ASObject>> GetEnumerator()
```
Comment (warriordog) [inline ActivityPub.Types/AS/Collection/ASCollection.cs:103]: Can this be simplified to pass through the iterator from `Items`? Something like this:
```suggestion
    public IEnumerator<Linkable<ASObject>> GetEnumerator()
        => Items?.GetEnumerator() ?? Enumerable.Empty<Linkable<ASObject>>();
```

File: ActivityPub.Types/AS/Collection/ASCollection.cs [hunk 2]
```diff
@@ -98,6 +99,32 @@ public int TotalItems
...
+    {
+        if (Items == null)
+        {
+            yield break;
+        }
+
+        foreach (var item in Items)
+        {
+            if (item.HasValue)
```
Comment (warriordog) [inline ActivityPub.Types/AS/Collection/ASCollection.cs:112]: If the previous change is not possible, then can we yield `item` directly? This version will churn memory by re-wrapping the link/value into a new Linkable.

File: ActivityPub.Types/AS/Collection/ASCollection.cs [hunk 3]
```diff
@@ -98,6 +99,32 @@ public int TotalItems
...
+            }
+
+            if (item.HasLink)
+            {
+                yield return item.Link;
+            }
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
```
Comment (warriordog) [inline ActivityPub.Types/AS/Collection/ASCollection.cs:124]: Would you mind switching this to expression syntax?
```suggestion
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
```

**Review by warriordog — CHANGES_REQUESTED** [review]
Thanks for the PR! I left a few comments on `ASCollection.cs`, but they also apply in `ASOrderedCollection.cs` as well. I just didn't want to duplicate them.

File: ActivityPub.Types/AS/Collection/ASCollection.cs (hunk 1 above)
Comment (CreggHancock) [reply ActivityPub.Types/AS/Collection/ASCollection.cs:103]: Yeah this should totally be doable! Great idea, thanks! Can't believe I didn't think of that

**Comment by warriordog** [comment]
@CreggHancock so sorry for the wait! this looks good.

**Review by warriordog — APPROVED** [review]

**Pull request merged**